### PR TITLE
feat(SPV-679): finalize contacts

### PR DESCRIPTION
**Pull request by chris-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided: [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet-web-frontend/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet-web-frontend/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: src/api/requests/contact.ts [hunk 1]
```diff
@@ -1,72 +1,50 @@
...
-    name: 'The Guy',
-    status: 'not-confirmed',
-  },
-];
+const encodeURIPaymail = (paymail: string) => {
+  // Remove control characters from the paymail
+  function* iterator() {
+    for (let i = 0; i < paymail.length; i++) {
+      const code = paymail.charCodeAt(i);
+      if (code > 32 && code !== 127) {
```
Comment (arkadiuszos4chain) [inline src/api/requests/contact.ts:43]: Nice one. But we should validate if it is a valid email address

File: src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx [hunk 2]
```diff
@@ -4,51 +4,74 @@ import { SrOnlySpan } from '@/styles';
...
+      setError('Wrong phone number.');
+      return;
+    }
+    if (!paymail || !name) {
+      setError('Both Paymail and Name are required');
+      return;
+    }
     setLoading(true);
-    setError(false);
+    setError(undefined);
```
Comment (arkadiuszos4chain) [inline src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx:55]: Just a question. Isn't it a good practice to not use `undefined` and use `null` instead?

Comment (chris-4chain) [reply src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx:55]: Just a reply 😄  Here I should have used `null` I think. I'll change this.

File: src/api/requests/contact.ts (hunk 1 above)
Comment (chris-4chain) [reply src/api/requests/contact.ts:43]: I had to add this function because I faced a strange behavior:
- When I copied an `paymail` from a list and pasted to the input and then try to send to the backend I got an error, sth like: "string includes control characters".

... So I decided to make sure that control characters will not be sent. 

> But we should validate if it is a valid email address

I'll add additional paymail check, in here: `ContactUpsertModal.tsx`

**Review by arkadiuszos4chain — APPROVED** [review]

File: src/components/ContactsList/_modals/ContactUpsertModal/ContactEditModal.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      onSubmitted={onSubmitted}
+      sucessMsg="Contact edited successfully!"
+      errorMsg="Error occurred while editing a contact."
+      modalTitle="Edit contact"
+      modalSubtitle=""
+      fields={fields}
+      disabledPaymailInput={true}
+    />
+  );
+};
```
Comment (Nazarii-4chain) [inline src/components/ContactsList/_modals/ContactUpsertModal/ContactEditModal.tsx:34]: If pass `contact` into `ContactUpsertModal` then it possible to move `useEffect` logic into `ContactUpsertModal`. Then call `useContactFields` in `ContactUpsertModal`

File: src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  errorMsg,
+  sucessMsg,
+  modalTitle,
+  modalSubtitle,
+  disabledPaymailInput,
+}) => {
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<string | null>(null);
+  const [successMessage, setSuccessMessage] = useState<string>('');
+  const { name, paymail, phone, setName, setPaymail, setPhone } = fields;
```
Comment (Nazarii-4chain) [inline src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx:38]: Add useEffect from `ContactEditModal` and call `useContactFields` here

File: src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  fields: ContactFields;
+  disabledPaymailInput: boolean;
+};
+
+export const ContactUpsertModal: FC<ContactUpsertModal> = ({
+  onSubmitted,
+  onCancel,
+  fields,
+  errorMsg,
+  sucessMsg,
```
Comment (Nazarii-4chain) [inline src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx:30]:
```suggestion
  successMsg,
```

File: src/api/requests/contact.ts [hunk 6]
```diff
@@ -1,72 +1,50 @@
...
+        yield code;
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply src/api/requests/contact.ts:47]: Response from AI (I agree with that)

## Option 1: Generator Function

Memory Efficiency: The main advantage of Option 1 is that it uses a generator function with a yield statement. Generators compute and return values on demand, avoiding the creation of a full intermediate array. This means they generally use less memory, especially for larger paymail strings.
No Intermediate Storage: Each character is processed, filtered, and potentially yielded without having to store the entire sanitized string in memory at once.

## Option 2: Array Methods

Intermediate Arrays: Option 2 uses array methods like Array.from(), filter(), and map(). Each of these methods creates a new intermediate array as it transforms the input string. This significantly increases temporary memory usage.
Less Efficient for Large Strings: The larger the input string, the more pronounced the memory difference is likely to be because more intermediate arrays are generated and held in memory before the final result is calculated.
Factors Affecting Memory Consumption

JavaScript Engine Optimizations: Modern JavaScript engines may optimize some array operations in Option 2, reducing memory impact. The actual difference might be less dramatic depending on the engine.
Length of the paymail string: The longer the string, the larger the intermediate arrays in Option 2 become, and therefore the larger the memory difference.
Garbage Collection: How quickly the JavaScript engine's garbage collector can reclaim the memory used by intermediate arrays in Option 2 will impact overall memory usage.
In Summary

For memory efficiency, **Option 1** (the generator function approach) is generally superior, especially when dealing with larger paymail strings. The extent of its advantage can depend on the factors mentioned above.

File: src/components/ContactsList/ContactsTable.tsx/ContactsTable.tsx [hunk 7]
```diff
@@ -101,7 +111,7 @@ export const ContactsTable: FC = () => {
             setJustAddedContact(false);
           }}
         >
-          {justAddedContact && contactForVerification.status === 'not-confirmed' && <JustAddedContactMsg />}
+          {justAddedContact && contactForVerification.status === 'unconfirmed' && <JustAddedContactMsg />}
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply src/components/ContactsList/ContactsTable.tsx/ContactsTable.tsx:114]: Right, thx 👍

File: src/components/ContactsList/_modals/ContactUpsertModal/ContactEditModal.tsx (hunk 3 above)
Comment (chris-4chain) [reply src/components/ContactsList/_modals/ContactUpsertModal/ContactEditModal.tsx:34]: No, because I applied here some kind of "compose strategy". 
The `ContactUpsertModal` is used also in `ContactAddModal` component.

File: src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx (hunk 4 above)
Comment (chris-4chain) [reply src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx:38]: Can't do that. See above comment 👆

File: src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx (hunk 5 above)
Comment (chris-4chain) [reply src/components/ContactsList/_modals/ContactUpsertModal/ContactUpsertModal.tsx:30]: Thanks :D

File: src/components/Input/PaymailAutocomplete.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    const contactsPaymails = useMemo(() => contacts?.map((c) => c.paymail) ?? [], [contacts]);
+
+    const isSelected = useMemo(
+      () => contactsPaymails.find((c) => c === paymailValue),
+      [contactsPaymails, paymailValue],
+    );
+
+    const { getRootProps, getInputProps, getListboxProps, getOptionProps, groupedOptions } = useAutocomplete({
+      id: 'paymail-autocomplete',
+      options: !isSelected && paymailValue != '' ? contactsPaymails : [],
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply src/components/Input/PaymailAutocomplete.tsx:23]: Not in this case. You're suggestion is not returning empty array when the condition is false.

**Review by Nazarii-4chain — APPROVED** [review]

**Comment by wregulski** [comment]
testing
